app/controllers/file_manager_controller.rb

```diff
@@ -0,0 +1,240 @@
+                filename: File.basename(absolute_path),
+                disposition: 'attachment')
+    end
+  end
+
+  def upload_file(path)
+    absolute_path = check_path_exist(path)
+    if absolute_path == BASE_DIRECTORY
+      flash[:error] = "You cannot upload files/create folders in the root directory " \
+        "#{view_context.link_to 'here', new_course_url, method: 'get'}" \
+        " if you want to create a new course."
+      flash[:html_safe] = true
+    else
+      raise ActionController::ForbiddenError unless File.directory?(absolute_path)
+
+      if check_instructor(absolute_path) && !params[:name].nil?
+        all_filenames = Dir.entries(absolute_path)
+        if params[:name] != ""
+          if all_filenames.include?(params[:name].to_s)
+            flash[:error] = "File/folder with name #{params[:name]} already exists"
+          else
+            # Creating a folder
+            dir = "#{absolute_path}/#{params[:name]}"
+            FileUtils.mkdir_p(dir)
+            flash[:success] = "Successfully created folder"
+          end
+        else
+          # Uploading a file
+          input_file = params[:file]
+          return unless input_file
+          if all_filenames.include?(input_file.original_filename)
+            flash[:error] = "File with name #{input_file.original_filename} already exists"
+          elsif input_file.size >= 1.gigabyte
+            flash[:error] = "File size is too large. Upload a file that is smaller than 1 GB."
+          else
+            File.open(Rails.root.join(absolute_path, input_file.original_filename), 'wb') do |file|
+              file.write(input_file.read)
+            end
+            flash[:success] = "Successfully uploaded file(s)"
+          end
+        end
+      end
+    end
+  end
+
+  def my_escape(string)
+    string.gsub(/([^ a-zA-Z0-9_.-]+)/) do
+      "%#{$1.unpack('H2' * $1.bytesize).join('%').upcase}"
+    end
+  end
+
+  def populate_directory(current_directory, current_url)
+    directory = Dir.entries(current_directory)
+    @directory = directory.map do |file|
+      abs_path_str = "#{current_directory}/#{file}"
+      stat = File.stat(abs_path_str)
+      is_file = stat.file?
+      if %w[. ..].include?(file)
+        inst = true
+        if current_directory == BASE_DIRECTORY
+          inst = false
+        end
+      else
+        abs_path = Pathname.new(abs_path_str)
+        inst = check_instructor(abs_path)
+      end
+      {
+        size: (if is_file
+                 begin
+                   stat.size
+                 rescue StandardError
+                   '-'
+                 end
+               else
+                 '-'
+               end),
+        type: (is_file ? :file : :directory),
+        date: begin
+                stat.mtime.strftime('%d %b %Y %H:%M')
+              rescue StandardError
+                '-'
+              end,
+        relative: my_escape("/file_manager/#{current_url}#{file}").gsub('%2F', '/'),
```

Comment:
Instead of gsubbing specific encodings, use something like `CGI.unescape` (if it is still necessary)

Reply:
Fixed.